Add XML serialization methods to Parameter class with factory-pattern handling

PauloGitHB opened the pull request:
### **Description**

#### **Context**

This PR is a proposal to introduce a serialization feature for Parameter objects in the parametertree module, allowing to do :

- The conversion of a parameter into an XML binary string..
- The conversion of an XML binary string into a parameter.

The addition of these methods simplifies saving, retrieving, and exchanging parameters, especially in projects like PyMoDAQ, where serialization is crucial for configuring and manipulating complex graphical interfaces or settings from instrument.

#### **Summary of Changes :**

- Integration of a Factory Pattern in parametertree/xml_factory.py :
  - Role : Generic interface to serialize hierarchical Parameter to XML strings and deserialize XML string to Parameter objects. This pattern introduces two classes: XMLParameter and XMLParameterFactory
  - XMLParameter: to be inherited by Parameter
        1. stored in the parametertree/xml_base.py module
        2. implements two methods to serialize/deserialize common options of a Parameter
        3. expose two abstract methods to serialize/deserialize their specific options: value, limits or other. Each object inheriting from Parameter should reimplement these specific abstract methods
  - XMLParameterFactory: to be used for serialization/deserialization usage
        1. stored in parametertree/xml_factory.py
        2. uses the already implemented registration mechanism (PARAM_TYPE dictionary) to infer which Parameter serialization methods to be called depending on their “type”
        3. implements general methods to serialize/deserialize a hierarchy of Parameters, using the registration mechanism to infer which object (and methods) should be called while going through the hierarchy
  - Avantage : Facilitates extensibility and simplifies maintenance to support new types of parameters in the future.
  - Adding methods in Parameter (through XMLParameter inheritance) for common options:
  - common_options_from_xml(el) :  
        Extracts common options (type, name, title, visible, etc.) from an XML element and returns them as a dictionary.
  - common_options_from_parameter() :  
        Converts the common options of a Parameter object into a dictionary to facilitate serialization.
- Addition of methods in custom Parameters (SimpleParameter as an example and then all others once you accepted our proposal) for type-specific options:
  - specific_options_from_xml(el) :  
        Interprets and converts specific options (value, limits, ...) of an XML element based on its type (bool, int, float, str, ...).
  - specific_options_from_parameter() :  
        Extracts and formats the value of a parameter to make it compatible with an XML representation.
- Addition of unit tests :
  - test_xml_parameter.py : Tests for generic and specific methods.
- Addition of a practical example :
  - A minimal working example (on SimpleParameters) has been added in the examples folder examples/parametertree_xml to illustrate the serialization and deserialization of parameters.

#### **Why is this important?**

This PR proposes a clear, standardized, and extensible management of parameters in XML format through a Factory Pattern. This allows in particular:

- A simplification of the serialization of parameters in tools like PyMoDAQ.
- A reduction in the complexity of the application-side code.

### **Tests added:**

### **Checklist :**

- Code tested locally
- Unit tests added
- Examples added to illustrate the new features
- Documentation updated for the new methods

PauloGitHubCopi committed:
no test

PauloGitHubCopi committed:
get specific set_specific

PauloGitHubCopi committed:
rename functions

PauloGitHubCopi committed:
test all passed plus doc string

PauloGitHubCopi committed:
basic options without test

PauloGitHubCopi committed:
test basic nad visible set true by default

PauloGitHubCopi committed:
factory plus specific limits ...

PauloGitHubCopi committed:
parametertree.py

PauloGitHubCopi committed:
test ok factory ok example to work

seb5g committed:
renamed or unused

seb5g committed:
split the XMLParameter and factory + cleaning + typing

seb5g committed:
inheritance from XMLParameter mixin

seb5g committed:
restored

seb5g committed:
implement XMLParameter abstract methods

seb5g committed:
restored

seb5g committed:
implement XMLParameter abstract methods

seb5g committed:
regitser numeric parameters

seb5g committed:
restored

seb5g committed:
xml factory minimal working example on simple parameters

seb5g committed:
restored

seb5g committed:
typing and docstring

seb5g committed:
cleaning

seb5g committed:
test serialize/deserialize

seb5g committed:
adding docstring

seb5g committed:
cleaned weird tests

PauloGitHB committed:
Merge pull request #1 from seb5g/pymodaq

Refactoring + typing + Cleaning + testing

pijyoi commented on pyqtgraph/parametertree/parameterTypes/bool.py:
```diff
@@ -1,5 +1,6 @@
 from ...Qt import QtWidgets
 from .basetypes import WidgetParameterItem
+from pyqtgraph.parametertree.parameterTypes.basetypes import WidgetParameterItem, SimpleParameter, Parameter
```
please do relative imports

PauloGitHB replied:
Ok i change it

pijyoi commented on pyqtgraph/parametertree/parameterTypes/color.py:
```diff
@@ -1,6 +1,7 @@
 from ... import functions as fn
 from ...widgets.ColorButton import ColorButton
 from .basetypes import SimpleParameter, WidgetParameterItem
+from qtpy import QtGui
```
please don't use qtpy, but in any case, QtGui doesn't seem to be needed?

PauloGitHB replied:
Ok I change it

pijyoi commented on pyqtgraph/parametertree/xml_base.py:
```diff
@@ -0,0 +1,84 @@
+from abc import abstractmethod
+from xml.etree import ElementTree as ET
+
+from pyqtgraph.parametertree import Parameter
```
please use relative imports

PauloGitHB replied:
Ok I change it

pijyoi commented on pyqtgraph/parametertree/parameterTypes/__init__.py:
```diff
@@ -36,9 +36,9 @@
 registerParameterType('colormap',  ColorMapParameter,    override=True)
 registerParameterType('file',      FileParameter,        override=True)
-registerParameterType('float',     SimpleParameter,      override=True)
+registerParameterType('float',     NumericParameter,      override=True)
```
There appears to be 4 additional types added by this PR (date, group, table, tableview) but not registered here. Was that intentional? This PR is quite large, perhaps it might be better to hold off those new types to another PR?

PauloGitHB replied:
Yes sorry it's a mistake caused by the bad push that I did

pijyoi commented on pyqtgraph/parametertree/parameterTypes/tableview.py:
```diff
@@ -0,0 +1,182 @@
+import importlib
+import json
+from qtpy import QtWidgets, QtCore
```
please don't use qtpy

PauloGitHB replied:
I delete this file

pijyoi commented on pyqtgraph/parametertree/parameterTypes/tableview.py:
```diff
@@ -0,0 +1,182 @@
+import json
+from qtpy import QtWidgets, QtCore
+from collections import OrderedDict
+from pyqtgraph.parametertree.Parameter import ParameterItem
```
please use relative import

PauloGitHB replied:
I delete this file

pijyoi commented on pyqtgraph/parametertree/parameterTypes/tableview.py:
```diff
@@ -0,0 +1,182 @@
+from .basetypes import Parameter
+from ..xml_parameter_factory import XMLParameterFactory, XMLParameter
+
+from pyqtgraph.widgets import ColorButton
```
please use relative import

PauloGitHB replied:
I delete this file

pijyoi commented on pyqtgraph/parametertree/parameterTypes/table.py:
```diff
@@ -0,0 +1,151 @@
+from qtpy import QtWidgets, QtCore
```
please don't use qtpy

PauloGitHB replied:
I delete this file

pijyoi commented on pyqtgraph/parametertree/parameterTypes/date.py:
```diff
@@ -0,0 +1,165 @@
+from qtpy import QtWidgets, QtCore
```
please don't use qtpy

PauloGitHB replied:
I delete this file

pijyoi commented on pyqtgraph/parametertree/parameterTypes/date.py:
```diff
@@ -0,0 +1,165 @@
+from qtpy import QtWidgets, QtCore
+from qtpy.QtCore import QDateTime, QTime
+from pyqtgraph import functions as fn
```
please use relative imports

PauloGitHB replied:
I delete this file

pijyoi commented on pyqtgraph/parametertree/parameterTypes/date.py:
```diff
@@ -0,0 +1,165 @@
+from pyqtgraph import functions as fn
+from .basetypes import WidgetParameterItem
+from .basetypes import SimpleParameter
+from pyqtgraph.parametertree import ParameterItem
```
please use relative imports

PauloGitHB replied:
I delete this file

pijyoi commented:
`XMLParameter` is defined as a Mix-In class. This PR inherits `XMLParameter` on the right. In my experience, mixin classes should be inherited on the left (in Python anyway). See #2662

PauloGitHubCopi committed:
clean for the PR

PauloGitHubCopi committed:
XML left inherit

PauloGitHB commented on tests/parametertree/test_Parameter.py:
```diff
@@ -16,8 +16,12 @@
 from pyqtgraph.parametertree.parameterTypes import GroupParameter as GP
 from pyqtgraph.Qt import QtGui
 
+from qtpy.QtCore import QDateTime
```
Ok i change it

PauloGitHB commented:
Hello, thank you for your comments. Indeed, this PR was not very clear because I made a mistake when pushing. The initial purpose of this PR was to focus only on SimpleParameter to see if it suits you, so that we can then implement these methods for other types of pyqtgraph, and perhaps also import new types. I have corrected all this so that this request only affects SimpleParameter. I am at your disposal if any other changes are necessary.

pijyoi commented on pyqtgraph/parametertree/parameterTypes/__init__.py:
```diff
@@ -16,7 +16,7 @@
 from .font import FontParameter, FontParameterItem
 from .list import ListParameter, ListParameterItem
-from .numeric import NumericParameterItem
+from .numeric import NumericParameterItem, NumericParameter
```
example doesn't run as NumericParameter does not exist

pijyoi commented on pyqtgraph/examples/parametertree.py:
```diff
@@ -149,5 +149,25 @@ def restore():
 
+
+
+## Create two ParameterTree widgets, both accessing the same data
```
The additional lines added here are almost a copy of the lines above. On my system, the example above doesn't display, probably because the variables used by the example below overwrite those above. But in any case, I am not sure what functionality the new example is supposed to demonstrate that isn't already demonstrated by the one above?

PauloGitHB replied:
In fact there was not supposed to be here because now our example is in parameter_xml.py so I just put the initial file.

pijyoi commented on pyqtgraph/examples/parametertree_xml.py:
```diff
@@ -0,0 +1,63 @@
+win = QtWidgets.QWidget()
+layout = QtWidgets.QGridLayout()
+win.setLayout(layout)
+layout.addWidget(QtWidgets.QLabel("These are two views of the same data. They should always display the same values."), 0,  0, 1, 2)
```
should change the label contents

pijyoi commented on pyqtgraph/examples/parametertree_xml.py:
```diff
@@ -0,0 +1,63 @@
+btn_export.clicked.connect(save_to_xml)
+btn_import.clicked.connect(load_from_xml)
+
+## Create two ParameterTree widgets, both accessing the same data
```
comment not true

pijyoi commented on pyqtgraph/examples/parametertree_xml.py:
```diff
@@ -0,0 +1,63 @@
+## Create two ParameterTree widgets, both accessing the same data
+t = ParameterTree()
+t.setParameters(p2, showTop=False)
+t.setWindowTitle('pyqtgraph example: Parameter Tree')
```
should change the title contents

pijyoi commented on pyqtgraph/examples/parametertree_xml.py:
```diff
@@ -0,0 +1,63 @@
+from pyqtgraph.Qt import QtWidgets
+
+app = pg.mkQApp("Parameter Tree Example")
+import pyqtgraph.parametertree.parameterTypes as pTypes
```
pTypes import not used

pijyoi commented on pyqtgraph/examples/parametertree_xml.py:
```diff
@@ -0,0 +1,63 @@
+import pyqtgraph.parametertree.parameterTypes as pTypes
+from pyqtgraph.parametertree import Parameter, ParameterTree
+
+from xml.etree import ElementTree as ET
```
ET import not used

pijyoi commented on tests/parametertree/test_parametertypes.py:
```diff
@@ -7,8 +7,10 @@
 from pyqtgraph.functions import eq
 from pyqtgraph.parametertree.parameterTypes import ChecklistParameterItem
 from pyqtgraph.Qt import QtCore, QtGui
+from pyqtgraph.parametertree.parameterTypes import SimpleParameter
```
SimpleParameter and Element imports are not used.
Probably should just revert the file to its original state.

PauloGitHB replied:
Yes I think so I just revert it

pijyoi commented on pyqtgraph/parametertree/xml_factory.py:
```diff
@@ -0,0 +1,150 @@
+
+            parent_xml_elt.append(elt)
+
+        return ET.tostring(parent_xml_elt)
```
This is out of my depth, but should ET.tostring be using the default encoding of "us-ascii" or something else?

seb5g replied:
well it is the default:
![image](https://github.com/user-attachments/assets/2a0bfa61-81c0-4778-a234-c7b9f410111b)

pijyoi commented:
I am just a simple user of the `ParameterTree` functionality in `pyqtgraph`.
But isn't there already save/restore functionality present?

```python
import pyqtgraph as pg
from pyqtgraph.Qt import QtWidgets
from pyqtgraph.parametertree import Parameter, ParameterTree

pg.mkQApp()

params = [
    {'title': 'Integer Parameter', 'name': 'param1', 'type': 'int', 'value': 10, },
    {'title': 'Float Parameter', 'name': 'param2', 'type': 'float', 'value': 3.14, },
    {'title': 'String Parameter', 'name': 'param3', 'type': 'str', 'value': 'Hello', },
    {'title': 'Boolean Parameter', 'name': 'param4', 'type': 'bool', 'value': True, }
]

params = Parameter.create(name='params', type='group', children=params)

saved_state = None

def save_to_something():
    global saved_state
    saved_state = params.saveState()
    print("Export:\n", saved_state)


def load_from_something():
    if saved_state:
        params.restoreState(saved_state)
        print("Import completed")

btn_export = QtWidgets.QPushButton("Export")
btn_import = QtWidgets.QPushButton("Import")
btn_export.clicked.connect(save_to_something)
btn_import.clicked.connect(load_from_something)

t = ParameterTree()
t.setParameters(params, showTop=False)

win = QtWidgets.QWidget()
layout = QtWidgets.QGridLayout()
win.setLayout(layout)
layout.addWidget(t, 0, 0, 1, 2)
layout.addWidget(btn_export, 1, 0)
layout.addWidget(btn_import, 1, 1)
win.show()

pg.exec()
```

seb5g commented:
> I am just a simple user of the `ParameterTree` functionality in `pyqtgraph`. But isn't there already save/restore functionality present?

This is not the same thing. We've developed a way to serialize a Parameter structure into a binary string. This allows the parameter structure to be saved as a text file but more importantly to be added in complex files accepting binary inputs. In the PyMoDAQ framework, we are saving numerical data in hdf5 files. However we are also saving metadata such as who, what, how and especially the contents of settings (instruments, configurations...) that are Parameter objects. This is where serialization is important. 

The saveState/restoreState is kind of the first step of this in the sense it transforms the objects in a dictionary hierarchy. Serialization from that dict could be simple (using json or xml) if Parameter objects contained only python builtins. However we have many Parameter whose value is not that simple to serialize: Colors, Date (and I introduced into pymodaq quite a few more that are complex)... The mechanism we propose allow to specify how specific options (value and extra options not shared by the base object) should be serialized.

PauloGitHubCopi committed:
PR clear parameter_xml adapted

PauloGitHubCopi committed:
remove useless spaces

PauloGitHubCopi committed:
remove useless import

pijyoi commented on pyqtgraph/parametertree/xml_base.py:
```diff
@@ -0,0 +1,84 @@
+            "removable": el.get('removable', '0') == '1',
+            "readonly": el.get('readonly', '0') == '1',
+            "tip": el.get('tip', ""),
+            "show_pb": el.get('show_pb', '0') == '1'
```
what is "show_pb" used for?

seb5g replied:
sorry it has been left here by mistake (it is related to a specific Parameter I will PR later)

pijyoi commented on pyqtgraph/parametertree/parameterTypes/basetypes.py:
```diff
@@ -283,6 +283,88 @@ def itemClass(self):
+        else:
+            raise TypeError(f'No interpreter found for type {el.get("type")}')
+        
+        for key in ["filetype", "detlist", "movelist", "addList", "addText", "label", "limits"]:
```
are all these attributes present for a SimpleParameter?

seb5g replied:
no you are right, only limits is useful!

seb5g replied:
the other ones are related to other types

PauloGitHubCopi committed:
remove show_pb and only limits for option

PauloGitHubCopi committed:
remove show pb from test

PauloGitHubCopi committed:
add static method and change options for parameter

PauloGitHB commented:
Hello! I hope you're doing well. I just wanted to follow up on my pull request to see if you had a chance to review it. I am more than willing to make any necessary modifications to facilitate the review process. Thank you very much for your time!

pijyoi commented:
Hello, I was only reviewing the syntactic aspects and am not really familiar with the parameter tree design. As mentioned, I am only a very simple user of parameter tree.

Hoping that the maintainers who are more familiar with the code can jump in.

PauloGitHB commented:
> Hello, I was only reviewing the syntactic aspects and am not really familiar with the parameter tree design. As mentioned, I am only a very simple user of parameter tree.
> 
> Hoping that the maintainers who are more familiar with the code can jump in.

PauloGitHB commented:
Thank you for taking the time to review the syntactic aspects of my pull request. I appreciate your insights and understand that the parameter tree design might require input from those more familiar with it. I hope that other maintainers can provide their feedback soon. Please let me know if there's anything else I can do in the meantime. Thanks again!

ntjess commented:
👋  Hi there, apologies for the late response to this thread... I'm more familiar with ParameterTree usage and can review the PR this weekend.

PauloGitHB commented:
Hi ,

I hope you're doing well! I just wanted to kindly follow up regarding the review of my PR. Let me know if there's anything you'd need from my side to facilitate the process. Looking forward to your feedback whenever you have a moment. Thank you, best regards!

ntjess commented:
I feel bad about the continual delays but my schedule has been so full lately 😅

I had a chance to pull locally and play with the example but not to read closely. I’ll finish it up this weekend (for real this time!) Thanks for your patience.

seb5g commented:
That's fine. Thanks

Le mer. 5 févr. 2025, 02:24, Nathan Jessurun ***@***.***> a
écrit :

> I feel bad about the continual delays but my schedule has been so full
> lately 😅
>
> I had a chance to pull locally and play with the example but not to read
> closely. I’ll finish it up this weekend (for real this time!) Thanks for
> your patience.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/pyqtgraph/pyqtgraph/pull/3203#issuecomment-2635495494>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABSNYI3DGDDZ4D7E5QF2RYD2OFR55AVCNFSM6AAAAABT5INEF6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMMZVGQ4TKNBZGQ>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

seb5g commented:
hello what's the status of our PR. Did you get the chance to read it?

ntjess committed:
Use more complex parameter tree for xml testing

ntjess commented:
@seb5g thanks for your patience, things have been a bit crazy for me. I completed an initial review and there are just a few styling changes.

I also integrated your serialization with the larger param tree example, which shows there are some types that are not supported. It would be good to have documentation (like a checklist) somewhere that indicates which types are supported vs. not. We don't need all types serializable yet since future PRs can extend the list, but it would be good for users to understand which types are supported out-of-the-box

Example: the current branch after my commit shows the error `TypeError: No interpreter found for type color`

j9ac9k commented:
Would say `NotImplementedError` would be the thing to raise.

https://docs.python.org/3/library/exceptions.html#NotImplementedError

seb5g commented:
Yes sure I will add a list and apply your correction. I actually have the serialization implemented for most types, I didn't want to have too much code for this first PR explaining the principles!

seb5g committed:
raise correct error in case serialization cannot occur

seb5g committed:
generate a report of unserializable yet Parameters

seb5g committed:
test serialization back and forth

error with QColor...

seb5g committed:
implement color for SimpleParameter

seb5g committed:
adding documentation section about xml serialization

seb5g committed:
removed generic xml serialization as it would fail because of not taken into account types

seb5g committed:
modified  parametertree example to generate a report of which Types are not covered by serialization

seb5g committed:
implement serialization of color type

seb5g committed:
refactor: rename, docstring and test for XML serialization implementation

seb5g committed:
Update xmlserialization.rst

seb5g committed:
Merge branch 'master' into master

seb5g committed:
ordering of options

seb5g committed:
add default handling

seb5g committed:
handling comparison of QColor of different address but same rgba color

seb5g committed:
something weird in dictionnary order comparison...

parameter.opts is not an Ordereddict but still for python >3.7 it was considered as such. But then savestate over two objects similar in all but options order failed!

seb5g commented:
Alright I implemented many things:

documentation
test
reports on serializable types
handling color type

Struggled quite a bit with QColor as with serialization the object is different (different address) but same value. I had to implement it in pyqtgraph.eq. But then another thing occured in the comparison related to order in the dictionnary parameter.opts... tested as an ordered dict in *eq* but is not...

Thanks for another round of review before I implement more types

j9ac9k commented on pyqtgraph/parametertree/parameterTypes/basetypes.py:
```diff
@@ -1,11 +1,15 @@
 import builtins
 
+from PyQt6.uic.Compiler.qtproxies import QtGui
```
You will need to import from pyqtgraph's own Qt module here, and we may need to add a shim to include the qtproxies (never used this module before).

j9ac9k commented on pyqtgraph/parametertree/parameterTypes/basetypes.py:
```diff
@@ -256,6 +260,27 @@ def eventFilter(self, obj, ev):
+    param_dict = {}
+    if attr in el.attrib:
+        value = el.get(attr)
+        if el.get('type') == 'bool':
```
nit-pick but wonder if this would be a suitable case to use the match/case syntax that's new in Python 3.10.

seb5g commented:
this PR is an early attempt for serialization superseeded by: https://github.com/pyqtgraph/pyqtgraph/pull/3444
